internal/tofu/transform_provider.go

```diff
@@ -102,147 +103,96 @@ func (t *ProviderTransformer) Transform(g *Graph) error {
 
 	var diags tfdiags.Diagnostics
 
-	// To start, we'll collect the _requested_ provider addresses for each
-	// node, which we'll then resolve (handling provider inheritance, etc) in
-	// the next step.
-	// Our "requested" map is from graph vertices to string representations of
-	// provider config addresses (for deduping) to requests.
-	type ProviderRequest struct {
-		Addr  addrs.AbsProviderConfig
-		Exact bool // If true, inheritance from parent modules is not attempted
-	}
-	requested := map[dag.Vertex]map[string]ProviderRequest{}
-	needConfigured := map[string]addrs.AbsProviderConfig{}
+	// Now we'll go through all the requested addresses we just collected and
+	// figure out which _actual_ config address each belongs to, after resolving
+	// for provider inheritance and passing.
+	m := providerVertexMap(g)
 	for _, v := range g.Vertices() {
```

Comment:
@cam72cam could you please explain why the [GraphNodeModulePath check](https://github.com/opentofu/opentofu/pull/2069/files#diff-2b903eb2dfe957cbd9f12f13c5d55dfc8ef718d021e97d3865e4cdf945a24f4dL181) is removed?

Reply:
The GraphNodeProviderConsumer interface is already required to be a GraphNodeModulePath (embedded).